module/zfs/dmu.c

```diff
@@ -81,6 +81,13 @@ int zfs_dmu_offset_next_sync = 0;
  */
 int zfs_object_remap_one_indirect_delay_ms = 0;
 
+/*
+ * Limit the amount we can prefetch with one call to this amount.  This
+ * helps to limit the amount of memory that can be used by prefetching.
+ * Larger objects should be prefetched a bit at a time.
+ */
+uint64_t dmu_prefetch_max = 8 * SPA_MAXBLOCKSIZE;
```

Comment:
Are the places in the code which need to be updated to now fetch larger objects in chunks?  For example, possibly `space_map_iterate()`.  We may also want to make this tuning available as a module option.

Reply:
In principle, yes.  However, in practice spacemaps don't get that big, so I'd rather keep this simple for now.

* dmu_prefetch_max is 128MB
* metaslabs are up to 16GB (for vdev sizes <=2PB)
* the worst case for a condensed spacemap is 1 entry per 2 sectors (alternating free and alloc'd sectors - this is way worse than ever seen in practice)
* uncondensed spacemap is at most zfs_condense_pct (2x) the condensed version
* spacemap entry size is 8 bytes

Therefore the max spacemap size is 16GB / (sectorsize * 2) * 2 * 8B = 256MB (for ashift=9) or 32MB (for ashift=12).

Reply:
> Is this tunable exported to userspace somehow, does it need to be documented?

It isn't, but it should be, thanks for mentioning that.